src/handler/mod.rs

```diff
@@ -1192,4 +1364,173 @@ impl Handler {
+                    Err(e) => {
+                        return Err(HolePunchError::Relay(e));
+                    }
+                };
+            self.send(tgt_node_address, packet).await;
+            Ok(())
+        } else {
+            // Either the session is being established or has expired. We simply drop the
+            // notification in this case to ensure hole punch round-trip time stays within the
+            // time out of the udp entrypoint for the target peer in the initiator's NAT, set by
+            // the original timed out FINDNODE request from the initiator, as the initiator may
+            // also be behind a NAT.
+            Err(HolePunchError::Relay(Discv5Error::SessionNotEstablished))
+        }
+    }
+
+    async fn new_connection(
+        &mut self,
+        enr: Enr,
+        socket_addr: SocketAddr,
+        conn_dir: ConnectionDirection,
+    ) {
+        if let Err(e) = self
+            .service_send
+            .send(HandlerOut::Established(enr, socket_addr, conn_dir))
+            .await
+        {
+            warn!(
+                "Failed to inform of established connection {}, {}",
+                conn_dir, e
+            )
+        }
+    }
+}
+
+#[async_trait]
+impl<P: ProtocolIdentity> NatHolePunch for Handler<P> {
+    async fn on_request_time_out(
+        &mut self,
+        relay: NodeAddress,
+        local_enr: Enr,
+        timed_out_nonce: MessageNonce,
+        target_session_index: NodeAddress,
+    ) -> Result<(), HolePunchError> {
+        // Another hole punch process with this target may have just completed.
+        if self.sessions.cache.get(&target_session_index).is_some() {
+            return Ok(());
+        }
+        if let Some(session) = self.sessions.cache.get_mut(&relay) {
+            let relay_init_notif =
+                Notification::RelayInit(local_enr, target_session_index.node_id, timed_out_nonce);
+            trace!(
+                "Sending notif to relay {}. relay init: {}",
+                relay.node_id,
+                relay_init_notif,
+            );
+            // Encrypt the message and send
+            let packet =
+                match session.encrypt_notification::<P>(self.node_id, &relay_init_notif.encode()) {
+                    Ok(packet) => packet,
+                    Err(e) => {
+                        return Err(HolePunchError::Initiator(e));
+                    }
+                };
+            self.send(relay, packet).await;
+        } else {
+            // Drop hole punch attempt with this relay, to ensure hole punch round-trip time stays
+            // within the time out of the udp entrypoint for the target peer in the initiator's
+            // router, set by the original timed out FINDNODE request from the initiator, as the
+            // initiator may also be behind a NAT.
+            warn!(
+                "Session is not established. Dropping relay notification for relay: {}",
```

Comment:
Line 1443 notif is used
Line 1462 notification is used
We should probably use notification for both